### PR TITLE
Validate clinical headers

### PR DESCRIPTION
**Pull request by fedde-s**

### CODE REVIEW

File: core/src/main/scripts/import_data_validator/validateData.py [hunk 1]
```diff
@@ -567,7 +566,10 @@ def validate(self):
                 return
 
             # parse the start-of-file comment lines
-            self.processTopLines(top_comments)
+            if not self.processTopLines(top_comments):
+                self.logger.info(
+                    'Invalid header comments, skipped data in file')
```
Comment (pieterlukasse) [inline core/src/main/scripts/import_data_validator/validateData.py:571]: what is meant by "skipped data"? Is the rest of the data not parsed because of this? If this is the case, it should be an error. Actually, in any case, this should be an error. If the header is invalid, the file should be considered invalid.

File: core/src/main/scripts/import_data_validator/validateData.py [hunk 2]
```diff
@@ -583,6 +585,8 @@ def validate(self):
                     '"' + dialect.delimiter in sample_content):
                 dialect.quoting = csv.QUOTE_NONE
             if not self._checkTsvDialect(dialect):
+                self.logger.info(
+                    'Invalid file format, skipped data in file')
```
Comment (pieterlukasse) [inline core/src/main/scripts/import_data_validator/validateData.py:589]: same comment here: should be an error.

File: core/src/main/scripts/import_data_validator/validateData.py (hunk 1 above)
Comment (pieterlukasse) [inline core/src/main/scripts/import_data_validator/validateData.py:571]: is this not related to the clinical data file? Maybe it should be moved to this specific validator instead.

Comment (fedde-s) [inline core/src/main/scripts/import_data_validator/validateData.py]: All the hook methods I added this notification to return whether they have
logged any errors that make the file unparseable.

File: core/src/main/scripts/import_data_validator/validateData.py [hunk 3]
```diff
@@ -622,8 +626,12 @@ def onComplete(self):
         self.logger.info('Validation of file complete')
 
     def processTopLines(self, line_list):
-        """Overide to parse any list of comment lines above the TSV header."""
-        pass
+        """Hook to parse any list of comment lines above the TSV header.
+
```
Comment (pieterlukasse) [inline core/src/main/scripts/import_data_validator/validateData.py]: same comment here. This is not something I would expect in the generic Validator class, but in ClinicalValidator

File: core/src/main/scripts/import_data_validator/validateData.py [hunk 4]
```diff
@@ -1220,24 +1231,165 @@ class ClinicalValidator(Validator):
...
+                if value in ('', 'NA'):
+                    self.logger.error(
+                        'Empty %s field in clinical attribute definition',
+                        LINE_NAMES[line_index],
+                        extra={'line_number': line_index + 1,
+                               'column_number': col_index + 1,
+                               'cause': value})
+                    invalid_values = True
+                if LINE_NAMES[line_index] in ('display_name', 'description'):
+                    pass
```
Comment (pieterlukasse) [inline core/src/main/scripts/import_data_validator/validateData.py:1299]: maybe I will find it in pr#2, but for the record: isn't this where the display_name and description values should be checked?

File: core/src/main/scripts/import_data_validator/validateData.py (hunk 1 above)
Comment (fedde-s) [inline core/src/main/scripts/import_data_validator/validateData.py:571]: All the hook methods I added this notification to return whether they have logged any errors that make the file unparseable, so this return value only serves to notify the caller of these errors.

Comment (fedde-s) [inline core/src/main/scripts/import_data_validator/validateData.py:571]: And top-of-file comments with certain metadata do occur in MAF files (and in other filetypes that may become supported in the future), but as far as I know cBioPortal currently ignores them, so they do not need to be validated. I still believe having a hook to do so could be useful though, so that Validator.validate() never needs to be overridden.

File: core/src/main/scripts/import_data_validator/validateData.py [hunk 5]
```diff
@@ -1220,24 +1231,165 @@ class ClinicalValidator(Validator):
...
+                    self.attr_defs[col_index]['attribute_type'].lower(),
+                    extra={'line_number': self.line_number,
+                           'column_number': col_index + 1,
+                           'cause': col_name})
+            else:
+                # copy the dict to make local changes
+                srv_attr = dict(srv_attr)
+                # compare defined values with existing ones
+                if 'attribute_type' in srv_attr:
+                    raise ValueError(
```
Comment (pieterlukasse) [inline core/src/main/scripts/import_data_validator/validateData.py:1373]: why a raise and not logger.error() ? Perhaps add a comment explaining why.

File: core/src/main/scripts/import_data_validator/validateData.py (hunk 4 above)
Comment (fedde-s) [inline core/src/main/scripts/import_data_validator/validateData.py:1299]: It would be, if there were any restrictions on their format other than the ones imposed by the tab-separated file format without a way of escaping tabs and newlines. Whether clinical attribute definitions correspond to the data already in the portal is checked in `ClinicalValidator.checkHeader()`, when parsing the column headers (the IDs of the attributes).

File: core/src/main/scripts/import_data_validator/validateData.py (hunk 5 above)
Comment (fedde-s) [inline core/src/main/scripts/import_data_validator/validateData.py:1373]: Will do, my comment there is indeed confusing. The reason is that this would not be an error in the staging files, but a change in the response of the server API that the validator script will have to be adapted to.

File: core/src/main/scripts/import_data_validator/validateData.py (hunk 4 above)
Comment (pieterlukasse) [inline core/src/main/scripts/import_data_validator/validateData.py:1299]: summary: add a comment like "validation of value against db is postponed to checkHeader" and perhaps just validate if the combination attribute_type + display_name is unique.

File: core/src/main/scripts/import_data_validator/validateData.py [hunk 6]
```diff
@@ -1253,6 +1405,38 @@ def checkLine(self, data):
...
+        '''Set cls.srv_sample_attrs and cls.srv_patient_attrs from portal.'''
+        if cls.srv_attrs is None:
+            cls.srv_attrs = cls._request_attrs(
+                SERVER_URL + '/api/clinicalattributes/patients', logger)
+            srv_sample_attrs = cls._request_attrs(
+                SERVER_URL + '/api/clinicalattributes/samples', logger)
+            id_overlap = (set(cls.srv_attrs.keys()) &
+                          set(srv_sample_attrs.keys()))
+            if id_overlap:
+                raise ValueError(
```
Comment (pieterlukasse) [inline core/src/main/scripts/import_data_validator/validateData.py:1419]: should not be an error here. If we want such a constraint, we should add it to the DB itself.

File: core/src/main/scripts/import_data_validator/validateData.py [hunk 7]
```diff
@@ -1485,30 +1669,36 @@ def processCaseListDirectory(caseListDir, cancerStudyId, logger):
 
     logger.info('Validation of case lists complete')
 
-def get_hugo_entrez_map(server_url):
+
+def request_from_portal_api(service_url, logger):
+    """Send a request to the portal API and return the decoded JSON object."""
+    if logger.isEnabledFor(logging.INFO):
```
Comment (pieterlukasse) [inline core/src/main/scripts/import_data_validator/validateData.py:1675]: why this if?

File: core/src/main/scripts/import_data_validator/validateData.py [hunk 8]
```diff
@@ -1632,7 +1820,7 @@ def main_validate(args):
         if logger.isEnabledFor(logging.ERROR):
```
Comment (pieterlukasse) [inline core/src/main/scripts/import_data_validator/validateData.py:1820]: remove this if?

File: core/src/test/scripts/unit_tests_validate_data.py [hunk 9]
```diff
@@ -76,14 +93,143 @@ def test_column_order_validation_ClinicalValidator(self):
...
         logger = getNewLogger()
-        logger.setLevel(logging.WARNING)
-        validator = ClinicalValidator(hugo_entrez_map,logger,meta_dict)
-        validator.validate()
-        # we expect no errors or warnings
-        self.assertEqual(0, len(logger.handlers[0].buffer))
-        # if the file has another order, this is also OK:
-        meta_dict['data_file_path'] = 'data_clin_order2.txt'
-        validator = ClinicalValidator(hugo_entrez_map,logger,meta_dict)
+        try:
```
Comment (pieterlukasse) [inline core/src/test/scripts/unit_tests_validate_data.py:96]: please align this with the new way you used in your tests and remove getNewLogger instead of keeping 2 implementations that do the same thing
